### PR TITLE
Feat: Client-side hooks - replace window.trigger

### PR DESCRIPTION
**Pull request by Antosik**
**Done**:
- All `$(window).trigger` entries were replaced by `hooks.fire`

### CODE REVIEW

**Comment by julianlam** [comment]
Oh! Wow, okay 😁

Do you want to work for NodeBB?

**Comment by barisusakli** [comment]
Replacing 
`$(window).trigger` with

```
require(['hooks'], function (hooks) {
   hooks.trigger()
});
```
Sometimes causes issues when the load order is important. So it's best to require hooks at the top of the file.

File: public/src/ajaxify.js [hunk 1]
```diff
@@ -449,9 +449,11 @@ $(document).ready(function () {
 					url: ev.state.returnPath,
 				}, ev.state.returnPath, config.relative_path + '/' + ev.state.returnPath);
 			} else if (ev.state.url !== undefined) {
-				ajaxify.go(ev.state.url, function () {
-					$(window).trigger('action:popstate', { url: ev.state.url });
-				}, true);
+				require(['hooks'], function (hooks) {
```
Comment (barisusakli) [inline public/src/ajaxify.js:452]: hooks is already defined at the top, no need to require here?

File: public/src/client/account/blocks.js [hunk 2]
```diff
@@ -54,7 +54,9 @@ define('forum/account/blocks', [
 					$('#users-container').html(html);
 					$('#users-container').siblings('div.alert')[html.length ? 'hide' : 'show']();
 				});
-				$(window).trigger('action:user.blocks.toggle', { data: payload });
+				require(['hooks'], function (hooks) {
```
Comment (barisusakli) [inline public/src/client/account/blocks.js:57]: suggest moving the require to the top

File: public/src/client/account/topics.js [hunk 3]
```diff
@@ -44,7 +44,9 @@ define('forum/account/topics', ['forum/account/header', 'forum/infinitescroll'],
 			html.find('.timeago').timeago();
 			app.createUserTooltips();
 			utils.makeNumbersHumanReadable(html.find('.human-readable-number'));
-			$(window).trigger('action:topics.loaded', { topics: topics });
+			require(['hooks'], function (hooks) {
```
Comment (barisusakli) [inline public/src/client/account/topics.js:47]: suggest moving to top of file

File: public/src/client/compose.js [hunk 4]
```diff
@@ -8,8 +8,10 @@ define('forum/compose', [], function () {
 		var container = $('.composer');
 
 		if (container.length) {
-			$(window).trigger('action:composer.enhance', {
-				container: container,
+			require(['hooks'], function (hooks) {
```
Comment (barisusakli) [inline public/src/client/compose.js:11]: suggest moving to top of file

File: public/src/ajaxify.js (hunk 1 above)
Comment (Antosik) [reply public/src/ajaxify.js:452]: Hooks at the top are defined in IIFE. This `document.ready` has no access to it.

**Comment by psychobunny** [comment]
nice :+1: I think this require can be removed? https://github.com/NodeBB/NodeBB/blob/42c164f878eaa794a298c752b8ab8fe04b9de720/public/src/app.js#L63-L71

File: public/src/ajaxify.js (hunk 1 above)
Comment (barisusakli) [reply public/src/ajaxify.js:452]: Ahh then we need another one in this IIFE, maybe right after `$(document).ready(function () {`

**Comment by Antosik** [comment]
> I think this require can be removed? (NodeBB/public/src/app.js#63-71)

I tried to remove it but it causes an error. `app.handleEarlyClicks` is called right after it was defined and `hooks` module isn't ready yet.

File: public/src/modules/translator.js [hunk 5]
```diff
@@ -9,7 +9,9 @@
 					namespace: namespace,
 					data: data,
 				};
-				$(window).trigger('action:translator.loadClient', payload);
+				require(['hooks'], function (hooks) {
```
Comment (barisusakli) [inline public/src/modules/translator.js:12]: this require needs to go up like in ajaxify.js

Comment (barisusakli) [reply public/src/modules/translator.js:12]: or gotto remove the resolve inside the require callback

**Pull request merged**